README.md

```diff
@@ -1,2 +1,105 @@
+
+**Procedure**: steps performed in a certain order to obtain a result.
+In the context of processing messages, when a Message of a certain type is received, a certain Procedure will process that Message.  
+One Procedure may process more than one type of Message. One Message may be processed by more than one Procedure.
+
+## Phases of message processing
+
+These are the Phases that a **Procedure Module** may contain:
+
+- **Message Creation Phase**  
+  In this phase the message with its payload is created.  
+  The **Procedure Module** may expose a utility function to ease the creation of the Message and its Payload that the Procedure processes.
+
+* **Validation Phase**  
+  In this pre-processing Phase the Procedure determine if the message can be processed or not.  
+  Some steps that may be performed in this Phase are:
+
+  - Validation: are the Message and its Payload valid?
+  - Authorization: is the user authorized to perform this Procedure?
+  - Ask the user for confirmation before proceeding
+
+* **Processing Phase**  
+  This is the main phase where the work is performed.  
+  The steps in this Phase are:
+
+  - Set the "processing" state
+  - Perform step(s)
+
+* **Processing Failed Phase**  
+  In this post-processing Phase the Procedure performs the steps needs to report and mitigate failures occurred during the Processing Phase.  
+  Possible steps in this Phase are:
+
+  - Clear the "processing" state
+  - Record error (Log, Track)
+  - May restore state to initial condition (Recover/Undo)
+  - May notify the user
+  - May retry by dispatching a retry Message
+
+* **Processing Succeeded Phase**  
+  In this post-processing Phase the Procedure performs the steps needs to record the success and the results of the Processing Phase.  
+  Possible steps in this Phase are:
+
+  - Clear the "processing" state
+  - Update the model
+  - May notify the user
+
+* **Broadcast Update Phase**  
+  In this last Phase the Procedure broadcasts a Message on what was updated. This may cause other Procedures to perform additional work.  
+  This Phase has a single step:
+
+  - Dispatch item updated message
+
+## Motivations for the "Procedure Module" pattern
+
+- The Phases of message processing are coupled. Changes to one Phase may affect the following Phase(s).
+- Placing all Phases of a message processing Procedure inside a Module increases the cohesion of the code.
+- Easily add new Procedures by adding new **Procedure Modules**, without the need to modify the existing ones.
+- It scales well in large teams where several people can work on separate **Procedure Modules** with no or minimal merge conflicts.
+
+## Analogy
+
+The **Organize Code by Features** pattern tells us to group all files that implement one feature under one ancestor folder.
+
+The **Procedure Module** pattern tells us to put all Phases of a message processing Procedure inside one code Module.  
+In this case the feature is the Procedure.
+
+## Testing
```

Comment:
It might be good to include a sentence explaining why black-box is recommended over white-box.

Reply:
Done:
Testing each Phase independently leads to implementation specific tests that have little value when refactoring code.